### PR TITLE
Allow building CUB tests without cuRand

### PR DESCRIPTION
**Pull request by senior-zero**
## Description

closes https://github.com/NVIDIA/cccl/issues/249

This PR allows building CUB tests without cuRand. The cuRand usage is disabled by default if there's no CTK found. To force this behavior, one can provide `-DCUB_C2H_ENABLE_CURAND=NO` CMake option.

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cub/test/CMakeLists.txt [hunk 1]
```diff
@@ -16,7 +16,14 @@ option(METAL_BUILD_EXAMPLES OFF)
...
 
-find_package(CUDAToolkit REQUIRED)
+find_package(CUDAToolkit)
+
+set(curand_default OFF)
+if (CUDAToolkit_FOUND)
+  set(curand_default ON)
+endif()
+
+option(CUB_C2H_ENABLE_CURAND "Use CUDA CURAND library" ${curand_default})
```
Comment (miscco) [inline cub/test/CMakeLists.txt:26]: Could that just be
```suggestion
option(CUB_C2H_ENABLE_CURAND "Use CUDA CURAND library" CUDAToolkit_FOUND)
```

File: cub/test/c2h/generators.cu [hunk 2]
```diff
@@ -36,12 +36,31 @@
 
 #include <c2h/custom_type.cuh>
 #include <c2h/generators.cuh>
-#include <curand.h>
 #include <fill_striped.cuh>
 
+#if C2H_HAS_CURAND 
+#include <curand.h>
+#else
```
Comment (miscco) [inline cub/test/c2h/generators.cu:43]: No Change requested: We recently tried to be more strict with our conditional compilation. If you dont mind you can change it to `else // ^^^ C2H_HAS_CURAND ^^^ / vvv !C2H_HAS_CURAND  vvv`

It is an complete overkill here, so feel free to ignore

File: cub/test/c2h/generators.cu [hunk 3]
```diff
@@ -125,28 +153,49 @@ RANDOM_TO_VEC_ITEM_SPEC(3, w);
...
+                   this->m_distribution.end(),
+                   i_to_rnd_t{m_re});
+  m_re.discard(this->m_distribution.size());
+#endif
+}
+
 float *generator_t::prepare_random_generator(seed_t seed,
                                              std::size_t num_items)
 {
+  m_distribution.resize(num_items);
```
Comment (miscco) [inline cub/test/c2h/generators.cu:190]: Should this be
```suggestion
   this->m_distribution.resize(num_items);
```

**Review by miscco — APPROVED** [review]

File: cub/test/c2h/generators.cu (hunk 2 above)
Comment (senior-zero) [reply cub/test/c2h/generators.cu:43]: Yeah, I'm trying not to introduce a text noise when it's not helpful. If it's a few lines, I'm omitting the annotation.

File: cub/test/c2h/generators.cu [hunk 4]
```diff
@@ -36,12 +36,31 @@
...
 
+#if !C2H_HAS_CURAND 
+struct i_to_rnd_t
+{
+  thrust::default_random_engine m_engine{};
+
+  template <typename IndexType>
+  __host__ __device__ float operator()(IndexType n)
+  {
+    m_engine.discard(n);
```
Comment (elstehle) [inline cub/test/c2h/generators.cu:58]: Just leaving this here for posterity. 
I was not sure about the complexity of `discard(n)` for thrust's default re. Per this comment that @senior-zero pointed me to, https://github.com/senior-zero/thrust_benchmark/blob/f193da487300b8534550dd15f2d3e86310062249/benches/tbm/range_generator.cuh#L145-L147 we believe it's `O(log(n))`. 
Hence, we decided to dismiss this complexity concern.

**Review by elstehle — APPROVED** [review]

**Review by wmaxey — APPROVED** [review]

**Pull request merged**